File: submissions/leopoldo-lima/solution/public/infrastructure/repositories/api-opportunity-repository.js
export class ApiOpportunityRepository {
  /**
   * @param {string} basePath
   */
  constructor(basePath = "/api") {
    this.basePath = basePath;
  }

  /**
   * @param {{
   *   region?: string,
   *   manager?: string,
   *   deal_stage?: string,
   *   q?: string,
   *   priority_band?: string,
   *   limit?: number
   * }} filters
   * @param {{signal?: AbortSignal}} options
   */
  async listOpportunities(filters = {}, options = {}) {
    const params = new URLSearchParams();
    if (filters.region) params.set("region", filters.region);
    if (filters.manager) params.set("manager", filters.manager);
    if (filters.deal_stage) params.set("deal_stage", filters.deal_stage);
    if (filters.q) params.set("q", filters.q);
    if (filters.priority_band) params.set("priority_band", filters.priority_band);
    params.set("limit", String(filters.limit ?? 20));

    const response = await fetch(`${this.basePath}/opportunities?${params.toString()}`, {
      signal: options.signal,
    });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return response.json();
  }

  /**
   * @param {string} id
   * @param {{signal?: AbortSignal}} options
   */
  async getOpportunity(id, options = {}) {
    const response = await fetch(`${this.basePath}/opportunities/${id}`, {
      signal: options.signal,
    });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return response.json();
  }

  /**
   * @param {{signal?: AbortSignal}} options
   */
  async getDashboardKpis(options = {}) {
    const response = await fetch(`${this.basePath}/dashboard/kpis`, {
      signal: options.signal,
    });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return response.json();
  }

  /**
   * @param {{signal?: AbortSignal}} options
   */
  async getDashboardFilterOptions(options = {}) {
    const response = await fetch(`${this.basePath}/dashboard/filter-options`, {
      signal: options.signal,
    });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return response.json();
  }
}

File: submissions/leopoldo-lima/solution/src/infrastructure/repositories/api_opportunity_repository.py
from __future__ import annotations

from dataclasses import dataclass

from src.api.contracts import (
    DashboardFilterOptionsResponse,
    DashboardKpisResponse,
    OpportunitiesListResponse,
    OpportunityDetailResponse,
)
from src.infrastructure.http.api_client import ApiClient
from src.infrastructure.http.errors import ApiClientError


class OpportunityRepositoryError(Exception):
    """Base repository error."""


class OpportunityNotFoundError(OpportunityRepositoryError):
    """Raised when opportunity ID is not found."""


@dataclass
class ApiOpportunityRepository:
    client: ApiClient

    def list_opportunities(
        self,
        region: str | None = None,
        manager: str | None = None,
        deal_stage: str | None = None,
        q: str | None = None,
        sort_by: str = "score",
        sort_order: str = "desc",
        limit: int = 20,
        page: int | None = None,
        page_size: int | None = None,
    ) -> OpportunitiesListResponse:
        try:
            return self.client.list_opportunities(
                region=region,
                manager=manager,
                deal_stage=deal_stage,
                q=q,
                sort_by=sort_by,
                sort_order=sort_order,
                limit=limit,
                page=page,
                page_size=page_size,
            )
        except ApiClientError as exc:
            raise OpportunityRepositoryError("Failed to list opportunities.") from exc

    def get_opportunity(self, opportunity_id: str) -> OpportunityDetailResponse:
        try:
            return self.client.get_opportunity(opportunity_id)
        except ApiClientError as exc:
            if getattr(exc, "status_code", None) == 404:
                raise OpportunityNotFoundError(
                    f"Opportunity '{opportunity_id}' not found."
                ) from exc
            raise OpportunityRepositoryError("Failed to fetch opportunity detail.") from exc

    def get_dashboard_kpis(self) -> DashboardKpisResponse:
        try:
            return self.client.get_dashboard_kpis()
        except ApiClientError as exc:
            raise OpportunityRepositoryError("Failed to fetch dashboard KPIs.") from exc

    def get_dashboard_filter_options(self) -> DashboardFilterOptionsResponse:
        try:
            return self.client.get_dashboard_filter_options()
        except ApiClientError as exc:
            raise OpportunityRepositoryError("Failed to fetch dashboard filters.") from exc
